### PR TITLE
Support BPartners without contract in commission deeds

### PR DESCRIPTION
**Pull request by cp-ps**
refs: https://github.com/metasfresh/metasfresh/issues/11488

### CODE REVIEW

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactory.java [hunk 1]
```diff
@@ -112,20 +129,66 @@ public CommissionConfigFactory(@NonNull final CommissionConfigStagingDataService
...
 				contractRequest.getCustomerBPartnerId(),
 				contractRequest.getSalesProductId());
 
 		return CollectionUtils.extractDistinctElements(contractId2Config.values(), Function.identity());
 	}
 
+	@NonNull
+	private ImmutableList<I_C_Flatrate_Term> createGenericContract(
+			@NonNull final I_C_Flatrate_Conditions flatrateCondition,
+			@NonNull final ImmutableList<I_C_Flatrate_Term> commissionTermRecords,
```
Comment (metas-ts) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactory.java:155]:
```suggestion
			@NonNull final ImmutableList<I_C_Flatrate_Term> existingCommissionTermRecords,
```

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactory.java [hunk 2]
```diff
@@ -112,20 +129,66 @@ public CommissionConfigFactory(@NonNull final CommissionConfigStagingDataService
...
+			@NonNull final ImmutableList<BPartnerId> hierarchyBPartnerIds,
+			@NonNull final Beneficiary endCustomer)
+	{
+		final ImmutableSet<Integer> salesRepIdsWithContract = commissionTermRecords.stream()
```
Comment (metas-ts) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactory.java:159]:
```suggestion
		final ImmutableSet<Integer> salesRepIdsWithContract = existingCommissionTermRecords.stream()
```

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactory.java [hunk 3]
```diff
@@ -112,20 +129,66 @@ public CommissionConfigFactory(@NonNull final CommissionConfigStagingDataService
...
 		final ImmutableMap<FlatrateTermId, CommissionConfig> contractId2Config = createCommissionConfigsFor(
-				commissionTermRecords,
+				commissionTermRecordsBuilder.build(),
 				contractRequest.getCustomerBPartnerId(),
 				contractRequest.getSalesProductId());
 
 		return CollectionUtils.extractDistinctElements(contractId2Config.values(), Function.identity());
 	}
 
+	@NonNull
```
Comment (metas-ts) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactory.java:152]:
```suggestion
	/**
	  * Creates commission contracts for bpartners that don't have one yet
	  */
	@NonNull
```

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactory.java [hunk 4]
```diff
@@ -112,20 +129,66 @@ public CommissionConfigFactory(@NonNull final CommissionConfigStagingDataService
...
+		final FlatrateTermCreator termCreator = FlatrateTermCreator.builder()
+				.ctx(Env.getCtx())
+				.products(ImmutableList.of(commissionProductRecord))
+				.startDate(Timestamp.from(Instant.now()))
+				.endDate(Timestamp.from(Instant.now().plus(NO_COMMISSION_AGREEMENT_DEFAULT_CONTRACT_DURATION)))
+				.conditions(flatrateCondition)
+				.bPartners(bPartnersWithoutContracts)
+				.isCompleteDocument(true)
+				.build();
+
```
Comment (metas-ts) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactory.java:188]: please log this to `Loggables.withLogger(Level.Info)..`

File: backend/de.metas.contracts/src/test/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactoryTest.java [hunk 5]
```diff
@@ -256,6 +276,211 @@ void createForNewCommissionInstances_multiple_contracts()
...
+
+		final CommissionConfig config = configs.get(0);
+
+		assertThat(config.getCommissionType()).isEqualTo(CommissionType.HIERARCHY_COMMISSION);
+
+		final HierarchyConfig hierarchyConfig = HierarchyConfig.cast(config);
+
+		final CommissionContract contractLvl1 = hierarchyConfig.getContractFor(Beneficiary.of(salesRepLvl1Id));
+		assertThat(contractLvl1).isNull();
+	}
```
Comment (metas-ts) [inline backend/de.metas.contracts/src/test/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactoryTest.java:482]: :tada:

**Review by metas-ts — CHANGES_REQUESTED** [review]

File: backend/de.metas.contracts/src/test/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactoryTest.java (hunk 5 above)
Comment (dm-ps) [reply backend/de.metas.contracts/src/test/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactoryTest.java:482]: 🎉🎉

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactory.java (hunk 4 above)
Comment (dm-ps) [reply backend/de.metas.contracts/src/main/java/de/metas/contracts/commission/commissioninstance/services/CommissionConfigFactory.java:193]: done

**Review by metas-ts — APPROVED** [review]

**Pull request merged**